feat: EPI-1184: Schedule Leave for a User (User Management Admin Panel)

namhoangbes opened the pull request:
### Changes

_Add a brief description of the changes in this PR to help give the reviewer context._

### Deploys

- [ ] **Deploy to Tamanu Internal** <!-- #deploy -->

### Tests

- [ ] **Run E2E Tests** <!-- #e2e -->

### Remember to...

- ...write or update tests
- ...add UI screenshots and **testing notes** to the Linear issue
- ...add any **manual upgrade steps** to the Linear issue
- ...update the [config reference](https://beyond-essential.slab.com/posts/reference-config-file-0c70ukly), [settings reference](https://beyond-essential.slab.com/posts/reference-settings-0blw1x2q), or any [relevant runbook(s)](https://beyond-essential.slab.com/topics/runbooks-bs04ml6c)
- ...call out additions or changes to **config files** for the deployment team to take note of

<!-- Thank you! -->

namhoangbes committed:
feat: EPI-1184: implement APIs to create and get user leaves

namhoangbes committed:
feat: EPI-1185: implement API to remove leave and enhance get user leaves

namhoangbes committed:
chore: update description for table user leaves

huytranbes committed:
Merge branch 'epic-scheduling-enhancements' of https://github.com/beyondessential/tamanu into feat/EPI-1184-schedule-leave-for-user

huytranbes committed:
Merge branch 'epic-scheduling-enhancements' of https://github.com/beyondessential/tamanu into feat/EPI-1184-schedule-leave-for-user

huytranbes committed:
feat(user-leave): add user leave management functionality for front-end

- Implemented `useCreateUserLeaveMutation` and `useDeleteUserLeaveMutation` for creating and deleting user leave requests.
- Added `useUserLeavesQuery` to fetch user leave data.
- Created `UserLeaveSection` component for scheduling and managing user leaves, integrated with the user profile modal.
- Enhanced user experience with toast notifications for success and error handling during leave operations.

huytranbes committed:
refactor(user-leave): simplify leave mutation logic and enhance UI components

- Removed unused `useDeleteUserLeaveMutation` and related query invalidation logic from `useCreateUserLeaveMutation`.
- Updated `UserLeaveSection` to improve state management for leave deletion and enhanced button styling.
- Refactored validation messages to utilize translation for better user experience.
- Adjusted layout of date fields to use a grid system for improved responsiveness.

huytranbes committed:
fix(user-leave): invalidate user leaves query after successful deletion

- Added query invalidation for user leaves in `UserLeaveSection` to ensure the UI reflects the latest state after a leave is removed.
- Enhanced user feedback with a success toast notification upon successful deletion.

huytranbes committed:
refactor(user-leave): clean up UserLeaveSection component by removing unused code

- Removed unused imports and state management related to leave deletion from `UserLeaveSection`.
- Simplified the component structure by eliminating the leave list rendering logic.
- Streamlined the validation schema and initial values for better clarity and maintainability.

huytranbes committed:
feat(users): EPI-1185: Manage Upcoming Scheduled Leave (User Management Admin Panel) (#8110)

feat(user-leave): enhance UserLeaveSection with leave deletion functionality

- Added leave deletion capability in `UserLeaveSection`, allowing users to remove scheduled leaves.
- Implemented confirmation modal for delete actions to improve user experience and prevent accidental deletions.
- Integrated leave list rendering to display upcoming scheduled leaves with formatted dates.
- Enhanced state management for deletion process and added success/error notifications for user feedback.

khoi-bes commented on packages/central-server/app/admin/users.js:
```diff
@@ -156,6 +162,118 @@ usersRouter.post(
+    const { startDate, endDate } = data;
+
+    const today = getCurrentDateString();
+    if (endDate < today) {
```
I think we need to parse this to Date to be able to compare as they are all string

khoi-bes commented on packages/database/src/models/UserLeave.ts:
```diff
@@ -0,0 +1,82 @@
+          allowNull: false,
+          field: 'user_id',
+        },
+        scheduledBy: {
```
please mimic other model for foreign key declaration, just for the consistency, we don't need to define it here

khoi-bes commented on packages/database/src/models/UserLeave.ts:
```diff
@@ -0,0 +1,82 @@
+  declare updatedAt: Date;
+  declare deletedAt?: Date;
+
+  static initModel({ primaryKey, ...options }: InitOptions) {
```
you should define the props (using declare) below this line (like other models)

namhoangbes committed:
fix(user-leave): update user leave model and migration

huytranbes committed:
feat(user-leave): implement leave deletion functionality and enhance user experience

- Added `useDeleteUserLeaveMutation` for deleting scheduled leaves.
- Updated `UserLeaveSection` to integrate leave deletion with confirmation modal and toast notifications for user feedback.
- Refactored leave creation and deletion logic for improved state management and user interaction.

huytranbes committed:
refactor(user-leave): replace date formatting function with utility function

- Removed the custom `formatDate` function from `UserLeaveSection` and replaced it with `formatShort` from `@tamanu/utils/dateTime` for improved date formatting consistency.
- Updated leave date rendering to utilize the new utility function, enhancing code clarity and maintainability.

huytranbes committed:
refactor(user-leave): clean up imports and enhance date formatting in UserLeaveSection

- Removed unused imports and streamlined the component structure in `UserLeaveSection`.
- Integrated `formatShort` utility for date formatting, improving consistency and clarity.
- Ensured that the component remains focused on its primary functionality while maintaining readability.
